client/kubernetes/hardware_finder.go

```diff
@@ -0,0 +1,173 @@
+package kubernetes
+
+import (
+	"context"
+	"net"
+
+	"github.com/packethost/pkg/log"
+	"github.com/pkg/errors"
+	"github.com/tinkerbell/boots/client"
+	"github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
+	"github.com/tinkerbell/tink/pkg/controllers"
+	"github.com/tinkerbell/tink/pkg/convert"
+	"github.com/tinkerbell/tink/protos/workflow"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// HardwareFinder is a type that looks up hardware from Kubernetes
+type HardwareFinder struct {
+	clientFunc func() crclient.Client
+	logger     log.Logger
+	namespace  string
+}
+
+// NewHardwareFinder returns a HardwareFinder that discovers hardware from Kubernetes.
+func NewHardwareFinder(logger log.Logger, k8sAPI, kubeconfig string) (*HardwareFinder, error) {
+	ccfg := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfig},
+		&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: k8sAPI}})
+
+	config, err := ccfg.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	namespace, _, err := ccfg.Namespace()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	manager, err := controllers.NewManager(config, controllers.GetServerOptions())
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	go manager.Start(context.Background())
```

Comment:
hmm this is surprising to me, guess my whole notion of k8s controllers is wrong here (**very** likely). Why does boots need to start a controller/manager to get data? Is it that boots will go through this locally started manager or something?

Reply:
hmmm...removing this goroutine from this function might be something to consider. The name suggests that its is a constructor and having a constructor start a Go routine seems a bit unusual. Also, Boots' [design philosophy](https://github.com/tinkerbell/boots/blob/main/docs/DESIGNPHILOSOPHY.md) encourages 
> "When you spawn goroutines, make it clear when - or whether - they exit. [2](https://github.com/golang/go/wiki/CodeReviewComments#goroutine-lifetimes)" 

additionally: https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions

Reply:
I think I see what you were going for here, keeping the same interface between tinkerbell and Kubernetes modes.

Passing the funcs back to main() would seem reasonable, and then have main() run things, or register the things that need running and let some runner handle the functions.

And then I looked at `main()` (with all its initialization and go funcs).  

> Boots could use a bit of a refactor/organization in how the backends work, but for now I've tried to fit in to the existing structure.

yep.

Reply:
The manager includes a client-side indexed cache and I was just re-using what is in Tink. Given that we don't need a controller/manager, we can probably just instantiate a [`cluster.Cluster()`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2/pkg/cluster#Cluster)